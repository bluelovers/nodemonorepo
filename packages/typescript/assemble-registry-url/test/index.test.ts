import createFactory, {
  FactoryParam,
  NPM_REGISTRY,
  YARN_REGISTRY,
  encodePackageName,
  validRegistry,
  validPackageName,
  validVersion
} from '../index'

describe('createFactory', () => {
  describe('when param is valid', () => {
    function nofn (param?: FactoryParam) {
      const base = createFactory(param)

      const result: {
        [_: string]: string
      } = {}

      for (const [key, value] of Object.entries(base)) {
        result[key] = typeof value === 'function' ? '[Function]' : value
      }

      return result
    }

    describe('when param is undefined', () => {
      it('should work', () => {
        expect(nofn()).toEqual(nofn({}))
      })
    })

    describe('when param.registry is undefined', () => {
      it('should use npm registry', () => {
        expect(nofn({})).toEqual(nofn({
          registry: NPM_REGISTRY
        }))
      })
    })

    describe('when param.registry is specified', () => {
      it('npm', () => {
        expect(createFactory({
          registry: NPM_REGISTRY
        })).toMatchSnapshot()
      })

      it('yarn', () => {
        expect(createFactory({
          registry: YARN_REGISTRY
        })).toMatchSnapshot()
      })
    })

    describe('when param.package is specified', () => {
      it('without param.registry', () => {
        expect(createFactory({
          package: 'package'
        })).toMatchSnapshot()
      })

      it('with param.registry', () => {
        expect(createFactory({
          registry: NPM_REGISTRY,
          package: 'package'
        })).toMatchSnapshot()
      })

      it('scoped', () => {
        expect(createFactory({
          package: '@scope/package'
        })).toMatchSnapshot()
      })
    })

    describe('when param.package and param.version is specified', () => {
      it('without param.registry', () => {
        expect(createFactory({
          package: 'package',
          version: '1.2.3'
        })).toMatchSnapshot()
      })

      it('with param.registry', () => {
        expect(createFactory({
          registry: NPM_REGISTRY,
          package: 'package',
          version: '1.2.3'
        })).toMatchSnapshot()
      })
    })
  })

  describe('when param is invalid', () => {
    it('because of param.registry', () => {
      expect(createFactory({
        registry: 'Invalid Registry!'
      })).toMatchSnapshot()
    })

    it('because of param.package', () => {
      expect(createFactory({
        package: '!@#$%^'
      })).toMatchSnapshot()
    })

    it('because of param.version', () => {
      expect(createFactory({
        package: 'package',
        version: 'Invalid Version Syntax!'
      })).toMatchSnapshot()
    })
  })
})

describe('Other functions', () => {
  function map<Right> (
    list: string[],
    fn: (x: string) => Right
  ): {
    readonly [_: string]: Right
  } {
    let result: {[_: string]: Right} = {}

    for (const item of list) {
      result[item] = fn(item)
    }

    return result
  }

  it('encodePackageName', () => {
    expect(map(
      [
        'package',
        '@scope/package'
      ],
      encodePackageName
    )).toMatchSnapshot()
  })

  it('validRegistry', () => {
    expect(map(
      [
        NPM_REGISTRY,
        YARN_REGISTRY,
        'Invalid Registry!'
      ],
      validRegistry
    )).toEqual({
      [NPM_REGISTRY]: true,
      [YARN_REGISTRY]: true,
      'Invalid Registry!': false
    })
  })

  it('validPackageName', () => {
    expect(map(
      [
        'package',
        '@scope/package',
        'Invalid Package Name!'
      ],
      validPackageName
    )).toEqual({
      'package': true,
      '@scope/package': true,
      'Invalid Package Name!': false
    })
  })

  it('validVersion', () => {
    expect(map(
      [
        '0.1.2',
        '1.2.3',
        'v3', // expect: invalid
        'Invalid Version!'
      ],
      validVersion
    )).toEqual({
      '0.1.2': true,
      '1.2.3': true,
      'v3': false,
      'Invalid Version!': false
    })
  })
})
