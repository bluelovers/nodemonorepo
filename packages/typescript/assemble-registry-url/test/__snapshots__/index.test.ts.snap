// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Other functions encodePackageName 1`] = `
Object {
  "@scope/package": "@scope%2Fpackage",
  "package": "package",
}
`;

exports[`createFactory when param is invalid because of param.package 1`] = `
Object {
  "package": "!@#$%^",
  "registry": "https://registry.npmjs.org",
  "registryURL": "https://registry.npmjs.org",
  "setPackage": [Function],
  "setRegistry": [Function],
  "validPackageName": false,
  "validRegistry": true,
}
`;

exports[`createFactory when param is invalid because of param.registry 1`] = `
Object {
  "registry": "Invalid Registry!",
  "setRegistry": [Function],
  "validRegistry": false,
}
`;

exports[`createFactory when param is invalid because of param.version 1`] = `
Object {
  "allVersionsURL": "https://registry.npmjs.org/package/versions",
  "latestVersionURL": "https://registry.npmjs.org/package/latest",
  "package": "package",
  "packageURL": "https://registry.npmjs.org/package",
  "registry": "https://registry.npmjs.org",
  "registryURL": "https://registry.npmjs.org",
  "setPackage": [Function],
  "setRegistry": [Function],
  "setVersion": [Function],
  "validPackageName": true,
  "validRegistry": true,
  "validVersion": false,
  "version": "Invalid Version Syntax!",
}
`;

exports[`createFactory when param is valid when param.package and param.version is specified with param.registry 1`] = `
Object {
  "allVersionsURL": "https://registry.npmjs.org/package/versions",
  "latestVersionURL": "https://registry.npmjs.org/package/latest",
  "package": "package",
  "packageURL": "https://registry.npmjs.org/package",
  "registry": "https://registry.npmjs.org",
  "registryURL": "https://registry.npmjs.org",
  "setPackage": [Function],
  "setRegistry": [Function],
  "setVersion": [Function],
  "validPackageName": true,
  "validRegistry": true,
  "validVersion": true,
  "version": "1.2.3",
  "versionURL": "https://registry.npmjs.org/package/1.2.3",
}
`;

exports[`createFactory when param is valid when param.package and param.version is specified without param.registry 1`] = `
Object {
  "allVersionsURL": "https://registry.npmjs.org/package/versions",
  "latestVersionURL": "https://registry.npmjs.org/package/latest",
  "package": "package",
  "packageURL": "https://registry.npmjs.org/package",
  "registry": "https://registry.npmjs.org",
  "registryURL": "https://registry.npmjs.org",
  "setPackage": [Function],
  "setRegistry": [Function],
  "setVersion": [Function],
  "validPackageName": true,
  "validRegistry": true,
  "validVersion": true,
  "version": "1.2.3",
  "versionURL": "https://registry.npmjs.org/package/1.2.3",
}
`;

exports[`createFactory when param is valid when param.package is specified scoped 1`] = `
Object {
  "allVersionsURL": "https://registry.npmjs.org/@scope%2Fpackage/versions",
  "latestVersionURL": "https://registry.npmjs.org/@scope%2Fpackage/latest",
  "package": "@scope/package",
  "packageURL": "https://registry.npmjs.org/@scope%2Fpackage",
  "registry": "https://registry.npmjs.org",
  "registryURL": "https://registry.npmjs.org",
  "setPackage": [Function],
  "setRegistry": [Function],
  "setVersion": [Function],
  "validPackageName": true,
  "validRegistry": true,
}
`;

exports[`createFactory when param is valid when param.package is specified with param.registry 1`] = `
Object {
  "allVersionsURL": "https://registry.npmjs.org/package/versions",
  "latestVersionURL": "https://registry.npmjs.org/package/latest",
  "package": "package",
  "packageURL": "https://registry.npmjs.org/package",
  "registry": "https://registry.npmjs.org",
  "registryURL": "https://registry.npmjs.org",
  "setPackage": [Function],
  "setRegistry": [Function],
  "setVersion": [Function],
  "validPackageName": true,
  "validRegistry": true,
}
`;

exports[`createFactory when param is valid when param.package is specified without param.registry 1`] = `
Object {
  "allVersionsURL": "https://registry.npmjs.org/package/versions",
  "latestVersionURL": "https://registry.npmjs.org/package/latest",
  "package": "package",
  "packageURL": "https://registry.npmjs.org/package",
  "registry": "https://registry.npmjs.org",
  "registryURL": "https://registry.npmjs.org",
  "setPackage": [Function],
  "setRegistry": [Function],
  "setVersion": [Function],
  "validPackageName": true,
  "validRegistry": true,
}
`;

exports[`createFactory when param is valid when param.registry is specified npm 1`] = `
Object {
  "registry": "https://registry.npmjs.org",
  "registryURL": "https://registry.npmjs.org",
  "setPackage": [Function],
  "setRegistry": [Function],
  "validRegistry": true,
}
`;

exports[`createFactory when param is valid when param.registry is specified yarn 1`] = `
Object {
  "registry": "https://registry.yarnpkg.com",
  "registryURL": "https://registry.yarnpkg.com",
  "setPackage": [Function],
  "setRegistry": [Function],
  "validRegistry": true,
}
`;
